support/node-health-check/src/main.rs

```diff
@@ -0,0 +1,107 @@
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+use std::{sync::Arc, time::Duration};
+
+use anyhow::Result;
+use reqwest::StatusCode;
+use serde::{Deserialize, Serialize};
+use structopt::StructOpt;
+use warp::Filter;
+
+/// Health struct returned by the RPC
+#[derive(Debug, PartialEq, Serialize, Deserialize, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct Health {
+    /// Number of connected peers
+    pub peers: usize,
+    /// Is the node syncing
+    pub is_syncing: bool,
+    /// Should this node have any peers
+    /// Might be false for local chains or when running without discovery.
+    pub should_have_peers: bool,
+}
+
+#[derive(Debug, StructOpt, Clone)]
+#[structopt(
+    name = "node-health-check",
+    about = "Sidecar for substrate node to report health compatible with AWS ALB."
+)]
+struct Opt {
+    /// Debug
+    #[structopt(short, long)]
+    debug: bool,
+
+    /// Set poll_interval
+    #[structopt(short = "i", default_value = "5")]
+    poll_interval: u64,
+
+    /// Set minimum peers
+    #[structopt(default_value = "2")]
+    min_peers: usize,
+
+    /// Node rpc endpoint
+    #[structopt(short, long, env, default_value = "http://localhost:9933/health")]
+    node_rpc_endpoint: String,
+
+    // The port to listen on
+    #[structopt(short, long, default_value = "9955")]
+    port: u16,
+}
+
+async fn poller(opts: Opt, healthy: Arc<AtomicBool>) {
+    loop {
+        let resp = async {
+            reqwest::get(&opts.node_rpc_endpoint)
+                .await?
+                .json::<Health>()
+                .await
+        };
+
+        match resp.await {
+            Ok(health) => {
+                let node_is_healthy = !health.is_syncing
+                    && (!health.should_have_peers || health.peers >= opts.min_peers);
+
+                if opts.debug {
+                    eprintln!("Response = {:?}", health);
+                    eprintln!("Healthy  = {:?}", node_is_healthy);
+                }
+                healthy.store(node_is_healthy, Ordering::Relaxed);
+            }
+            Err(error) => {
+                eprintln!("{:?}", error);
+                healthy.store(false, Ordering::Relaxed);
+            }
+        }
+
+        tokio::time::sleep(Duration::from_secs(opts.poll_interval)).await;
+    }
+}
+
+async fn return_health_status(healthy: Arc<AtomicBool>) -> Result<impl warp::Reply, Infallible> {
+    if healthy.load(Ordering::Relaxed) {
+        Ok(StatusCode::OK)
+    } else {
+        Ok(StatusCode::NOT_FOUND)
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    let opts = Opt::from_args();
+    let healthy = Arc::new(AtomicBool::new(true));
+
+    let health_check = {
+        let healthy = healthy.clone();
+        warp::path!("health")
+            .map(move || healthy.clone())
+            .and_then(return_health_status)
+    };
+
+    tokio::spawn(poller(opts.clone(), healthy));
+
+    warp::serve(health_check)
+        .run(([0, 0, 0, 0], opts.port))
```

Comment:
If you start warp before the poller, you can avoid cloning the opts. Alternatively, pass a reference to the opts.

Reply:
Obviously performance doesn't matter, but it's a smell to have a pile of clones and copies unnecessarily throughout the code.

Reply:
Good point, I did this the first way around. I also do not like too many `clones`. It just does not feel good. In this case we do the await on `warp::serve` and the program blocks. We can do it, but then we have to `tokio::spawm` the `warp::serve` as well. This would avoid the clone. In this case I decided to go for the extra clone. Considering its not a time critical loop. But we think the same way on this point.

Reply:
I fixed in an easier way, just doing the blocking await on the poller.